### ISSUE TITLE
Variable usage

### ISSUE THREAD

**Issue by jas-** [h4l/json.bash#10]
Thanks for putting this together. I am hoping to move some existing functions to leverage this library, however the syntax is proving to be difficult when using variables to create a complex json object.

```sh
foo=bar
obj={\"foo\":\"${foo}\"}
jb @obj:json
```

Perhaps I am missing something obvious?

**Comment by h4l** [comment]
Hey, you're welcome, thanks for using it and your question.

The problem here is that the `obj` variable is not exported by your shell. Because `jb` runs json.bash in a new bash shell, the obj variable is not visible unless it's exported. If I run this myself I see:

```console
$ foo=bar
$ obj={\"foo\":\"${foo}\"}
$ jb @obj:json
json(): Could not process argument '@obj:json'. Its value references unbound variable $obj. (Use the '~' flag after the :type to treat a missing value as empty.)
␘
```

I assume you're writing a shell script in bash yourself (rather than using the shell interactively). In this case the best thing to do is to use the `json` function instead of `jb`. The difference is that `json` is just a bash function, so it runs in the same bash process as the rest of your code and can see the bash variables without exporting them. Whereas `jb` runs a new bash process that itself calls `json`, so it can only see variables that the calling process exports.

So you should be able to do this:
```shell
source json.bash  # need to import the functions to use them in your script

foo=bar
obj={\"foo\":\"${foo}\"}
json @obj:json  # outputs: {"obj":{"foo":"bar"}}
```

If you were doing it interactively, you can either explicitly export obj, or define obj while running jb:

```console
$ foo=bar
$ obj={\"foo\":\"${foo}\"}

$ obj=${obj:?} jb @obj:json
{"obj":{"foo":"bar"}}

$ export obj
$ jb @obj:json
{"obj":{"foo":"bar"}}
```

---

As an aside, if you're writing a script and would like to have the best performance, you can use the `out` argument of the `json` function to have it encode into a variable rather than printing to stdout. This allows you to capture the output without capturing stdout (which often requires forking a subshell):

```shell
source json.bash
foo=bar
out=obj json @foo  # this encodes into the $obj variable
json @obj:json  # prints: {"obj":{"foo":"bar"}}
```

A tip is to use `declare -p` to inspect bash variables, e.g. here we can see it created an array;

```console
$ declare -p obj
declare -a obj=([0]="{\"foo\":\"bar\"}")
```

You could append another JSON value to the array by using out=obj again:

```console
$ out=obj json test:number=123

$ # now we have two values in obj (I should probably have renamed the var 🙃)
$ declare -p obj
declare -a obj=([0]="{\"foo\":\"bar\"}" [1]="{\"test\":123}")

$ # So we can create an array with the values it holds
$ json @obj:json[]
{"obj":[{"foo":"bar"},{"test":123}]}

$ # or flatten them as properties
$ json ...@obj:json{:json}
{"foo":"bar","test":123}
```

And if you don't mind forking (not a big issue unless you're doing something thousands of times), you can simplify it by using process substitution to nest json calls:
```console
$ json obj:json@<(json foo=bar)
{"obj":{"foo":"bar"}}
```

---

Hope that helps, let me know if you have trouble.

**Comment by jas-** [comment]
Thanks for the help! I am seeing some odd behavior if I try a complex example. See here.

```sh
read -r hostname os version kernel arch <<< "$(env_set_env)"
out=system json @hostname @os @version @kernel @arch
declare -p system
echo "${system[@]}"
json @system:json
```

While the `declare` and `echo` of the exported array reflects properly, using the `json @system:json` throws an error... complete output below.

```sh
declare -a system=([0]="{\"hostname\":\"solaris\",\"os\":\"Solaris\",\"version\":\"11\",\"kernel\":\"11.4.42.111.0\",\"arch\":\"i386\"}")
{"hostname":"solaris","os":"Solaris","version":"11","kernel":"11.4.42.111.0","arch":"i386"}
bash: json_validator_PID: parameter null or not set
[1]   Exit 2                  coproc json_validator ( LC_ALL=C.UTF-8 grep --only-matching --line-buffered -P -e "${validation_request//[' 
']/}" )
```

**Comment by h4l** [comment]
That's odd, I can confirm that shouldn't happen! If I try this I get the correct output without the error.

It's an error to do with the JSON validation. To validate JSON, it starts a single long-lived grep process in the background and feeds it lines of JSON to validate. This grep process is started using bash's coprocess feature, and that sets a `_PID` var after creating the coprocess; it's this var that is not set for you.

I wonder if coprocess is missing or somehow disabled in your version of bash. Could you share the `bash --version` output? And also the `grep --version` too for good measure. (The tests run against several versions, the earliest is [4.4.19](https://github.com/h4l/json.bash/blob/main/docker-bake.hcl#L32).)

As a workaround, you could replace `:json` for `:raw` which will include the JSON string as `:json` does, except that it does no validation. (Not ideal for a real situation, but would at least unblock you.)

**Comment by jas-** [comment]
Well it is `5.1.12(1)-release (x86_64-pc-solaris2.11)`. If I move to using `json @system:raw` it works.

`{"system":{"hostname":"solaris","os":"Solaris","version":"11","kernel":"11.4.42.111.0","arch":"i386"}}`

Let me see if I see the same results on Red Hat. I have found the bash distribution for Solaris can sometimes have quirks.

**Comment by h4l** [comment]
I just tried running a solaris VM to get an idea what's going on. First thing I notice is that the provided grep program is not compatible with the flags json.bash is using. json.bash is expecting grep to be GNU Grep. Specifically, it runs grep with these flags: `grep --only-matching --line-buffered -P -e ...`.

So I think what's happening is that the coprocess exits due to grep not supporting the flags, resulting in the `_PID` var not being set. If you could install GNU Grep and have that be available at `grep` then I think it would work. I think it defaults to ggrep rather than grep though. I should probably have json.bash detect this error condition and show a sensible error message.

I tried to see if I could install ggrep on my solaris VM, but my knowledge of Solaris is basically 0! Not sure if my VM is borked, but it couldn't search for packages:
```console
opc@instance-20240720-0954:~$ pkg search grep
Failed to set locale: unsupported locale setting.  Falling back to C.
pkg: Unable to set locale en_US.UTF-8; locale package may be broken or
not installed.  Reverting to C locale.
pkg: Some repositories failed to respond appropriately:
solaris:
  Framework error: code: E_RECV_ERROR (56) reason: OpenSSL SSL_read: OpenSSL/3.0.13: error:0A00045C:SSL routines::tlsv13 alert certificate required, errno 0
URL: 'https://pkg.oracle.com/solaris/support/versions/0/'
```

**Comment by h4l** [comment]
I guess grep compatibility is another good reason I should rewrite json.bash in a proper language (for the `jb` command) and perhaps implement the `json` bash API using a coprocess to call to the native version...

**Comment by jas-** [comment]
I like the fact you are using `bash` quite frankly. The project I have been working on is strictly bash so this fits perfectly.

The error you are seeing is because you don't have a publisher of packages enabled. See `man pkg publisher` or `pkg help publisher`. 

Anyways, provided the `json:raw` seems to work for this use case I think having a conditional on the OS will suffice.

I do have one more question but will open a new issue for it. Thanks again for your help on this!

**Comment by h4l** [comment]
Yeah, there's definitely value in having a bash-only solution. Maybe two implementations is the optimal approach. Thanks for the Solaris tip!
